File: Scala/day8.scala
import java.nio.file.{Files, Paths}
import scala.io.Source
import org.scalatest._
import flatspec._
import matchers._

class Day8Solver {

    def parseData(input: Array[String]) : Array[Array[Byte]] = Array.tabulate(input.length, input.length)((row, col) => (input(row)(col) - '0').toByte)

    def getRow(trees: Array[Array[Byte]], row: Int) : Seq[Byte] = 0 to trees.length - 1 map(col => trees(row)(col))
    def getCol(trees: Array[Array[Byte]], col: Int) : Seq[Byte] = 0 to trees.length - 1 map(row => trees(row)(col))

    // First, determine whether there is enough tree cover here to keep a tree house hidden. To do this, you need to count the
    // number of trees that are visible from outside the grid when looking directly along a row or column.
    // A tree is visible if all of the other trees between it and an edge of the grid are shorter than it. Only consider trees in
    // the same row or column; that is, only look up, down, left, or right from any given tree.
    // Puzzle == Consider your map; how many trees are visible from outside the grid?
    def puzzle1(trees: Array[Array[Byte]]) : Int = (1 to trees.length - 2).flatMap(row => (1 to trees.length - 2).map(col => (row, col)))
                                                                          .filter((row, col) => visible(trees, trees(row)(col), row, col))
                                                                          .length +
                                                    trees.length * 2 * 2 - 4

    def visible(trees: Array[Array[Byte]], tree: Byte, row: Int, col: Int) : Boolean =
        getRow(trees, row).take(col).forall(c => c < tree) ||
        getRow(trees, row).drop(col + 1).forall(c => c < tree) ||
        getCol(trees, col).take(row).forall(r => r < tree) ||
        getCol(trees, col).drop(row + 1).forall(r => r < tree)

    // The Elves just need to know the best spot to build their tree house: they would like to be able to see a lot of trees.
    // To measure the viewing distance from a given tree, look up, down, left, and right from that tree; stop if you reach an
    // edge or at the first tree that is the same height or taller than the tree under consideration. (so a tree on the edge == 0)
    // A tree's scenic score is found by multiplying together its viewing distance in each of the four directions.
    // Puzzle == What is the highest scenic score possible for any tree?
    def puzzle2(trees: Array[Array[Byte]]) : Int = (0 to trees.length - 1).flatMap(row => (0 to trees.length - 1).map(col => (row, col)))
                                                                          .map((row, col) => scenicScore(trees, trees(row)(col), row, col))
                                                                          .max

    def scenicScore(trees: Array[Array[Byte]], tree: Byte, row: Int, col: Int) : Int =
        viewingDistance(tree, getRow(trees, row).take(col).reverse) *
        viewingDistance(tree, getRow(trees, row).drop(col + 1))     *
        viewingDistance(tree, getCol(trees, col).take(row).reverse) *
        viewingDistance(tree, getCol(trees, col).drop(row + 1))

    def viewingDistance(tree: Byte, lineOfSight: Seq[Byte]) : Int = {
        val los = lineOfSight.takeWhile(t => t < tree).length
        if(los < lineOfSight.length) {
            return los + 1;
        }
        else {
            return los;
        }
    }
}

class Day8 extends AnyFlatSpec with should.Matchers {

    var sampleData = Array(
        "30373",
        "25512",
        "65332",
        "33549",
        "35390",
    )

    "Puzzle 1" should "count all trees visible from outside the forest in the sample data" in {
        val day8 = new Day8Solver
        var trees = day8.parseData(sampleData);
        day8.puzzle1(trees) should be (21)
    }

    "Puzzle 2" should "find the tree with the most scenic score in the sample data" in {
        val day8 = new Day8Solver
        var trees = day8.parseData(sampleData);
        day8.puzzle2(trees) should be (8)
    }

    val realData = Source.fromFile(new java.io.File(new java.io.File(".").getCanonicalPath).getParent() + "/.input/day8.data")
                         .getLines
                         .toArray

    "Puzzle 1" should "count all trees visible from outside the forest in the AoC data" in {
        val day8 = new Day8Solver
        var trees = day8.parseData(realData);
        day8.puzzle1(trees) should be (1717)
    }

    "Puzzle 2" should "find the tree with the most scenic score in the AoC data" in {
        val day8 = new Day8Solver
        var trees = day8.parseData(realData);
        day8.puzzle2(trees) should be (321975)
    }
}

File: CSharp/day8.cs
namespace AdventOfCode2022;

using NUnit.Framework;
using FluentAssertions;

using matthiasffm.Common.Math;

[TestFixture]
public class Day08
{
    private byte[,] ParseData(string[] lines)
    {
        var trees = new byte[lines.Length, lines[0].Length];
        trees.Populate((r, c) => (byte)(lines[r][c] - '0'));
        return trees;
    }

    [Test]
    public void TestSamples()
    {
        var data = new [] {
            "30373",
            "25512",
            "65332",
            "33549",
            "35390",
        };
        var trees = ParseData(data);

        Puzzle1(trees).Should().Be(21);
        Puzzle2(trees).Should().Be(2 * 2 * 1 * 2);
    }

    [Test]
    public void TestAocInput()
    {
        var data = File.ReadAllLines(@"day8.data");
        var trees = ParseData(data);

        Puzzle1(trees).Should().Be(1717);
        Puzzle2(trees).Should().Be(321975);
    }

    // First, determine whether there is enough tree cover here to keep a tree house hidden. To do this, you need to count the
    // number of trees that are visible from outside the grid when looking directly along a row or column.
    // A tree is visible if all of the other trees between it and an edge of the grid are shorter than it. Only consider trees in
    // the same row or column; that is, only look up, down, left, or right from any given tree.
    // Puzzle == Consider your map; how many trees are visible from outside the grid?
    private int Puzzle1(byte[,] trees)
    {
        // TODO: too slow O(n*n*n)
        //       precalc 4 lookop matrices which contain the rolling max for every direction on every position in O(n*n)
        return trees.Select((t, r, c) => IsVisible(trees, t, r, c))
                    .Sum();
    }

    private int IsVisible(byte[,] trees, byte tree, int row, int col) =>
        row == 0 || col == 0 || row == trees.GetLength(0) - 1 || col == trees.GetLength(1) - 1 ||
        trees.Row(row).Take(col).All(c => c.Item2 < tree) ||
        trees.Row(row).Skip(col + 1).All(c => c.Item2 < tree) ||
        trees.Col(col).Take(row).All(r => r.Item2 < tree) ||
        trees.Col(col).Skip(row + 1).All(r => r.Item2 < tree)
        ?
        1 : 0;

    // The Elves just need to know the best spot to build their tree house: they would like to be able to see a lot of trees.
    // To measure the viewing distance from a given tree, look up, down, left, and right from that tree; stop if you reach an
    // edge or at the first tree that is the same height or taller than the tree under consideration. (so a tree on the edge == 0)
    // A tree's scenic score is found by multiplying together its viewing distance in each of the four directions.
    // Puzzle == What is the highest scenic score possible for any tree?
    private int Puzzle2(byte[,] trees)
    {
        // TODO: too slow O(n*n*n)
        //       precalc 4 lookop matrices which contain the rolling lenght of local downramps for every direction on every position in O(n*n)
        return trees.Select((t, r, c) => ScenicScore(trees, t, r, c))
                    .Max();
    }

    private int ScenicScore(byte[,] trees, byte tree, int row, int col) =>
        ViewingDistance(tree, trees.Row(row).Take(col).Reverse()) *
        ViewingDistance(tree, trees.Row(row).Skip(col + 1)) *
        ViewingDistance(tree, trees.Col(col).Take(row).Reverse()) *
        ViewingDistance(tree, trees.Col(col).Skip(row + 1));

    private int ViewingDistance(byte tree, IEnumerable<(int, byte)> lineOfSight)
    {
        var scenicScore = 0;

        foreach(var loS in lineOfSight)
        {
            scenicScore++;

            if(loS.Item2 >= tree)
            {
                break;
            }
        }

        return scenicScore;
    }
}
